### ISSUE TITLE
Tray Icon & Title Missing

### ISSUE THREAD

**Issue by andersonjwan** [xournalpp/xournalpp#1442]
**Affects versions :**
 - OS: [Linux - Fedora: 5.2.8-200.fc30.x86_64]
 - Version of Xournal++ [v1.0.13, Built on Aug 23, 2019]

**Describe the bug**
Starting a new instance of Xournal++ works as intended; however, in the GNOME taskbar, the tray icon and name is missing. The results are shown in the screenshots below.

**To Reproduce**
1. Install all required Fedora dependencies as listed.
2. Clone the repository.
3. Compile.
4. Install system-wide from build directory directly.

**Expected behavior**
The tray icon is expected to be on the left of the title. The title should be 'Xournal++' and not 'xournalpp'.

**Screenshots of Problem**
![image](https://user-images.githubusercontent.com/52432873/63629378-a1303b80-c5c6-11e9-8305-e5da166f8582.png)
![image](https://user-images.githubusercontent.com/52432873/63629431-fbc99780-c5c6-11e9-8cea-eb0b20b7ed30.png)

**Additional context**
Xournal++ does show up in the Activities bar, so it must be just a tray icon issue of finding the title and image. Also, please disregard the date of the second screenshot - this was done on a snapshot of a VM. It was indeed built on Aug 23, 2019.

If this is a duplicate issue, please let me know. I can take this down if it has been patched, but not pushed.

**Comment by Technius** [comment]
In #1414, we changed the name of the xournalpp desktop and icon files to `com.github.xournalpp.xournalpp`. If you "installed" xournalpp instead of using a tarball release, you  will need to "reinstall" these files by renaming them (you could also delete them and run the install command again).

For reference, here are the files that need to be renamed (replace `xournalpp` with `com.github.xournalpp.xournalpp`):
```plain
${CMAKE_INSTALL_PREFIX}/share/icons/hicolor/scalable/apps/xournalpp.svg
${CMAKE_INSTALL_PREFIX}/share/mime/packages/xournalpp.xml
${CMAKE_INSTALL_PREFIX}/share/applications/xournalpp.desktop
${CMAKE_INSTALL_PREFIX}/share/thumbnailers/xournalpp.thumbnailers
${CMAKE_INSTALL_PREFIX}/share/metainfo/xournalpp.appdata.xml
```

**Comment by andersonjwan** [comment]
I completed a fresh installation from scratch as done before with a DE that has never had Xournal++ installed, and I do indeed see these specific file names updated in the installation output as shown below. Additionally, I verified that the files in each location were renamed and this also was true; however, the issue still persists.

`-- Installing: /usr/local/share/icons/hicolor/scalable/apps/com.github.xournalpp.xournalpp.svg`
`-- Installing: /usr/local/share/mime/packages/com.github.xournalpp.xournalpp.xml`
`-- Installing: /usr/local/share/applications/com.github.xournalpp.xournalpp.desktop`
`-- Installing: /usr/local/share/thumbnailers/com.github.xournalpp.xournalpp.thumbnailer`
`-- Installing: /usr/local/share/metainfo/com.github.xournalpp.xournalpp.appdata.xml`

I also ran `sudo ./cmake/postinst configure` and was returned with:
`gtk-update-icon-cache: Cache file created successfully.`

When Xournal++ is first started, the tray icon and title both appear with the rotating circling to indicate its loading, but once it finishes loading and is ready, the resulting image appears as screenshotted in the initial post.

**Comment by Technius** [comment]
If you have an old desktop file lying around, it might be confusing the DE into using the wrong icon path (at least from what I've tested); please also see my edit to my previous comment. I'll try testing on a machine with no Xournal++ installation tomorrow if deleting the old desktop files doesn't fix the issue.

**Comment by andersonjwan** [comment]
I think I have fixed it actually! Thanks to your help, I went into my `applications` folder located under `usr/local/share/applications` and noticed the file was actually labeled `com.github.xournalpp.xournalpp.desktop`; and so, simply renaming the file to `xournalpp.desktop` corrected the problem! I'm not too sure why the desktop file was built with that name, but using the original name fixed the problem, and the tray icon and title are correctly showing up! Shall I close this thread?

**Comment by lehmanju** [comment]
The file name is intended as this is a naming convention used by Gnome and KDE. However, the same issue appears when running NextcloudClient. Might be a Gnome issue.

**Comment by lehmanju** [comment]
```
diff --git a/desktop/com.github.xournalpp.xournalpp.appdata.xml b/desktop/com.github.xournalpp.xournalpp.appdata.xml
index fb0d4c2b..60ee9622 100644
--- a/desktop/com.github.xournalpp.xournalpp.appdata.xml
+++ b/desktop/com.github.xournalpp.xournalpp.appdata.xml
@@ -1,7 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!-- Copyright 2019 Tobias Mueller <tobiasmue@gnome.org>  -->
 <component type="desktop">
-  <id>com.github.xournalpp.xournalpp</id>
+  <id>com.github.xournalpp.xournalpp.desktop</id>
   <translation type="gettext">xournalpp</translation>
   <name>Xournal++</name>
   <summary>Take handwritten notes</summary>
@@ -63,9 +62,8 @@
       <image>https://i.imgur.com/3uxu40d.png</image>
     </screenshot>
   </screenshots>
-  <launchable type="desktop-id">com.github.xournalpp.xournalpp.desktop</launchable>
   <provides>
-    <id>xournalpp.desktop</id>
+    <binary>xournalpp</binary>
   </provides>
   <update_contact>webmaster_AT_huberulrich.de</update_contact>
   <content_rating type="oars-1.0"/>
diff --git a/desktop/com.github.xournalpp.xournalpp.desktop b/desktop/com.github.xournalpp.xournalpp.desktop
index 95de8678..95d071c8 100644
--- a/desktop/com.github.xournalpp.xournalpp.desktop
+++ b/desktop/com.github.xournalpp.xournalpp.desktop
@@ -1,7 +1,6 @@
 [Desktop Entry]
 Encoding=UTF-8
 Type=Application
-
 Name=Xournal++
 Comment=Take handwritten notes
 Name[fr]=Xournal++
@@ -18,10 +17,9 @@ Name[zh_CN]=Xournal++
 Comment[zh_CN]=手写注释工具，可为PDF文件添加各种元素的注释标记
 Name[ru]=Xournal++
 Comment[ru]=Делайте рукописные заметки
-
 Exec=xournalpp %f
+Icon=com.github.xournalpp.xournalpp
 Terminal=false
 StartupNotify=true
 MimeType=application/x-xoj;application/x-xojpp;application/x-xopp;application/x-xopt;application/pdf;
-Icon=com.github.xournalpp.xournalpp
 Categories=Office;GNOME;GTK;
```
This at least fixed the missing icon on dashboard.

**Comment by TheAssassin** [comment]
A bit off topic, but I've been looking into this as well. I use the AppImage, and despite the files being installed correctly by our integration code:

```sh
> xdg-mime query default application/x-xopp
appimagekit_43c5a5d0157a403ad0fe630d1bdd4b80-Xournal++.desktop
```

I'm still trying to find out why Xfce doesn't make use of this. `file --mime` and `xdg-mime query filetype some.xopp` both yield `application/gzip`. I've updated the local MIME db, then it worked, so I guess we need to that command upon integration.


Anyway, I'd like to suggest adding some proprietary magic bytes. If you look at the file header, there's some blank bytes that might be used:

```sh
> hexdump -Cv some.xopp | head -n1
00000000  1f 8b 08 00 00 00 00 00  00 03 bc bd 6b 53 1d bb  |............kS..|
```

There's the standard magic bytes for gzip, 0x1f, 0x8b, 0x08. As you can see, there's a lot of null bytes where you might embed a custom string that will clearly mark files as Xournal++ files:

```sh
> echo -ne 'xopp' | dd of=test.xopp bs=1 count=4 seek=4 conv=notrunc
4+0 records in
4+0 records out
4 bytes copied, 0.000208423 s, 19.2 kB/s
> hexdump -Cv test.xopp | head -n1
00000000  1f 8b 08 00 78 6f 70 70  00 03 bc bd 6b 53 1d bb  |....xopp....kS..|
```

Works fine, and is much more robust than just file extension guessing. And it's really easy to note this down in your MIME definition.

**Comment by lehmanju** [comment]
@muelli this has been introduced after your commit (https://github.com/xournalpp/xournalpp/commit/f7248ab1e4082c1627f333b3b97143bda3ba6da7), do you have any ideas?

**Comment by muelli** [comment]
hm. looks like a severe bug.
But I can't reproduce.

This is what it looks for me:

![image](https://user-images.githubusercontent.com/322807/64131534-81202980-cdc9-11e9-971f-1605d41aea65.png)

![image](https://user-images.githubusercontent.com/322807/64131541-8c735500-cdc9-11e9-92d2-9f4d4c3bc4d6.png)


I have used this to test: `flatpak install https://dl.flathub.org/build-repo/7039/com.github.xournalpp.xournalpp.flatpakref`.
I don't have xournalpp installed otherwise.

The patch in https://github.com/xournalpp/xournalpp/issues/1442#issuecomment-526550356 looks entirely wrong.

Replying to https://github.com/xournalpp/xournalpp/issues/1442#issuecomment-524573092
@andersonjwan can you install to `/usr/` rather than `/usr/local`? 
Can you try the flatpak to see whether it works for you?

**Comment by muelli** [comment]
> I have used this to test: `flatpak install https://dl.flathub.org/build-repo/7039/com.github.xournalpp.xournalpp.flatpakref`.

I should have added that this is based on https://github.com/flathub/com.github.xournalpp.xournalpp/commit/dd8bb3758d47a2c31cce0eb59c0502795556528d which in turn builds 7a728c6453fd7bbb0d2caadc4bf992aaaba3365f which hopefully has all the suspected changes incorporated.

**Comment by TheAssassin** [comment]
Has anyone tried to `appstreamcli validate <myfile>`? Seems fine to me...

**Comment by Technius** [comment]
@TheAssassin The appdata file on `master` (6260dba) fails to validate for me:
```plain
$ appstream-util validate desktop/com.github.xournalpp.xournalpp.appdata.xml 
desktop/com.github.xournalpp.xournalpp.appdata.xml: FAILED:
• markup-invalid        : <id> does not have correct extension for kind
Validation of files failed
```
#1452 updates the appdata file and should fix the validation problem. I'll need to test a further to see if the PR can also fix this issue.

**Comment by muelli** [comment]
FWIW: The validation for flathub works fine:

```
$ flatpak run org.freedesktop.appstream-glib validate desktop/*.appdata.xml
desktop/com.github.xournalpp.xournalpp.appdata.xml: O
$ flatpak run org.freedesktop.appstream-glib --version
Version:	0.7.16
$
```

A locally installed version is also fine with the file as is:

```
$ appstream-util  validate desktop/*.appdata.xml
desktop/com.github.xournalpp.xournalpp.appdata.xml: OK
$ appstream-util --version
Version:	0.7.7
$
```

For completeness, this is the file I am validating: 
```
35684af452ecad3e997dd57e4770e438e58a2355893bb188fc8fb6e896c33d59  desktop/com.github.xournalpp.xournalpp.appdata.xml
```

**Comment by Technius** [comment]
It might be that the version of `appstream-glib` packaged in the distro I'm using (Solus Linux) is outdated, which is why I'm getting a validation error.

**Comment by TheAssassin** [comment]
> It might be that the version of appstream-glib packaged in the distro I'm using (Solus Linux) is outdated, which is why I'm getting a validation error.

That's the big issue with AppStream and the reason people find it hard to work with. The format seems to be neither stable nor backwards compatible...

**Comment by muelli** [comment]
> That's the big issue with AppStream and the reason people find it hard to work with. The format seems to be neither stable nor backwards compatible...

Hm.  That's interesting. I bet @ximion wants to hear opinions on that. Do you have references reg. people claiming the format being hard to work with or that stability is a problem?

**Comment by ximion** [comment]
> > That's the big issue with AppStream and the reason people find it hard to work with. The format seems to be neither stable nor backwards compatible...
> 
> Hm. That's interesting. I bet @ximion wants to hear opinions on that. Do you have references reg. people claiming the format being hard to work with or that stability is a problem?

We only ever broke compatibility once in 2014 when the `<application/>` tag was depecated and `<id/>` tags got values that were more freely defined. Apart from that instance, we never broke backward compatibility.
Code should work with older metainfo files just as well as it does with newer ones. However, the validators became more refined, which means newer validators may not validate something that the older ones did.
The `appstreamcli` tool of the reference implementation has different issues severities to communicate how big of an issue the reported problem actually is. Only errors and warnings fail validation, and it will take a bunch of releases before a deprecated feature jumps from being just an info message to being a validation error.
The `appstream-util` of appstream-glib doesn't have different severities per-issue, and can be really strict when run with `validate` as parameter. In that case you should simply use `validate-relax` to make style checks non-fatal.
So tl;dr, the specification is very stable, validators are constantly refined, so ideally use the newest versions of the validators or at least pay attention to `appstreamcli`'s issue severities (not everything makes validation fail, some thing may be ignored) and try `validate-relax` when validating with `appstream-util`.
FWIW, from quickly looking through the issue, this problem here looks unrelated to the metainfo file and more like an issue with the desktop-entry file.

**Comment by TheAssassin** [comment]
> Do you have references reg. people claiming the format being hard to work with or that stability is a problem?

https://github.com/AppImage/AppImageKit/search?q=appstream&type=Issues

I'm only describing our users' experiences with this validation. The format may be stable, okay, but the validation of it surely isn't. My impression is warnings are added and removed in different versions inconsistently. Using the latest version isn't always an option. Different versions of the same OS can yield new issues. That's pretty annoying when really working with this if you as a developer are new to working with AppData and your build system and workstation system differ.

What I really miss is an option to only fail validation on warnings by the way. Warnings are not errors. Sure, I want them printed out, but I get annoyed if they're treated as errors. But that's off topic. Yeah, I know, I should write an issue over at the repository.

**Comment by ximion** [comment]
> > Do you have references reg. people claiming the format being hard to work with or that stability is a problem?
> 
> https://github.com/AppImage/AppImageKit/search?q=appstream&type=Issues
> 
> I'm only describing our users' experiences with this validation. The format may be stable, okay, but the validation of it surely isn't. My impression is warnings are added and removed in different versions inconsistently. Using the latest version isn't always an option. Different versions of the same OS can yield new issues. That's pretty annoying when really working with this if you as a developer are new to working with AppData and your build system and workstation system differ.

That impression comes from there being two implementations of AppStream, unfortunately.
If you validate with only one, the experience should be better.

> What I really miss is an option to only fail validation on warnings by the way. Warnings are not errors. Sure, I want them printed out, but I get annoyed if they're treated as errors. But that's off topic. Yeah, I know, I should write an issue over at the repository.

Warnings are serious enough to make validations fail. In `appstreamcli` a warning basically means "this is not completely broken, but we may loose metadata due to it or the result may be unintended". That said, `appstreamcli` will allow you to silence any issue that isn't an error in future versions and recently received some refactoring to make forward-compatibility easier to achieve.
Maybe give it a try!

**Comment by TheAssassin** [comment]
I think we should just bundle `appstreamcli` whenever we want to validate AppStream data then. Is there e.g., an issue for me to track with regards to the changes you mentioned?

**Comment by ximion** [comment]
> I think we should just bundle `appstreamcli` whenever we want to validate AppStream data then. Is there e.g., an issue for me to track with regards to the changes you mentioned?

The refactored validator was released with AppStream 0.12.8, as for ignoring issue tags that's currently on my TODO list, but with a big question mark (people shouldn't want to ignore issues, so I am currently waiting for an actual usecase before implementing that feature - it is potentially not needed with the recent validator).

**Comment by LittleHuba** [comment]
Okay this discussion went way off-topic...
So to sumarize the current state of the issue:
- Renaming the desktop file to the old name fixes the icon not showing up but we actually want to use the new name -> We need to try to find out why the new name is not accepted by the DE and how to make it accept the change
- We should use magic bytes for the filetype -> Please create a new issue for that @TheAssassin 
- Appstream messes up the validation of our appdata file -> Create a new issue where we can discuss the steps we will take @TheAssassin

**Comment by lehmanju** [comment]
This **only** happens on GNOME/Wayland. Icon shows fine on GNOME/Xorg, KDE, etc.

(As the Nextcloud client application had the same issue I tested it with all WMs as well and it shows the same behavior.)

**Comment by Technius** [comment]
Do you know the cause? And does the workaround here (`make uninstall` with older version) work?

**Comment by lehmanju** [comment]
This has been fixed in Nextcloud now https://github.com/nextcloud/desktop/pull/1449
